core/src/telemetry/tracing.ts

```diff
@@ -0,0 +1,389 @@
+    // Setting empty llm request and response (as UI expect these) while not
+    // applicable for tool_response.
+    'gcp.vertex.agent.llm_request': '{}',
+    'gcp.vertex.agent.llm_response': '{}',
+  });
+
+  let functionResponseEventJson: string;
+  try {
+    functionResponseEventJson = JSON.stringify(functionResponseEvent, (key, value) => {
+      if (value === undefined) {
+        return undefined;
+      }
+      return value;
+    });
+  } catch (error) {
+    functionResponseEventJson = '<not serializable>';
+  }
+
+  span.setAttribute('gcp.vertex.agent.tool_response', functionResponseEventJson);
+}
+
+/**
+ * Traces a call to the LLM.
+ *
+ * This function records details about the LLM request and response as
+ * attributes on the current OpenTelemetry span.
+ *
+ * @param invocationContext The invocation context for the current agent run.
+ * @param eventId The ID of the event.
+ * @param llmRequest The LLM request object.
+ * @param llmResponse The LLM response object.
+ */
+export function traceCallLlm(
+  invocationContext: InvocationContext,
+  eventId: string,
+  llmRequest: LlmRequest,
+  llmResponse: LlmResponse
+): void {
+  // Special standard Open Telemetry GenAI attributes that indicate
+  // that this is a span related to a Generative AI system.
+  const span = trace.getActiveSpan();
+  if (!span) return;
+  
+  span.setAttributes({
+    'gen_ai.system': 'gcp.vertex.agent',
+    'gen_ai.request.model': llmRequest.model,
+    'gcp.vertex.agent.invocation_id': invocationContext.invocationId,
+    'gcp.vertex.agent.session_id': invocationContext.session.id,
+    'gcp.vertex.agent.event_id': eventId,
+    // Consider removing once GenAI SDK provides a way to record this info.
+    'gcp.vertex.agent.llm_request': safeJsonSerialize(buildLlmRequestForTrace(llmRequest)),
+  });
+
+  // Consider removing once GenAI SDK provides a way to record this info.
+  if (llmRequest.config) {
+    if (llmRequest.config.topP !== undefined) {
+      span.setAttribute('gen_ai.request.top_p', llmRequest.config.topP);
+    }
+    if (llmRequest.config.maxOutputTokens !== undefined) {
+      span.setAttribute('gen_ai.request.max_tokens', llmRequest.config.maxOutputTokens);
+    }
+  }
+
+  let llmResponseJson: string;
+  try {
+    // TypeScript doesn't have model_dump_json method, so we use JSON.stringify
+    llmResponseJson = JSON.stringify(llmResponse, (key, value) => {
+      if (value === undefined) {
+        return undefined;
+      }
+      return value;
+    });
+  } catch (error) {
+    llmResponseJson = '<not serializable>';
+  }
+
+  span.setAttribute('gcp.vertex.agent.llm_response', llmResponseJson);
+
+  if (llmResponse.usageMetadata) {
+    span.setAttribute('gen_ai.usage.input_tokens', llmResponse.usageMetadata.promptTokenCount || 0);
+    if (llmResponse.usageMetadata.candidatesTokenCount !== undefined) {
+      span.setAttribute('gen_ai.usage.output_tokens', llmResponse.usageMetadata.candidatesTokenCount);
+    }
+  }
```

Comment:
you can reduce if nesting to improve readability:

```suggestion
  if (llmResponse.usageMetadata) {
    span.setAttribute('gen_ai.usage.input_tokens', llmResponse.usageMetadata.promptTokenCount || 0);
  }
  
 if (llmResponse.usageMetadata?.candidatesTokenCount) {
    span.setAttribute('gen_ai.usage.output_tokens', llmResponse.usageMetadata.candidatesTokenCount);
  }
```